### DISCUSSION TITLE
Connecting to Databricks

### DISCUSSION THREAD

**Discussion by hbergmey** [slick/slick#2540] [Questions] [answered]
How do you connect to a Databricks DB? I have a connection string, an authentication token, and the Databricks JDBC driver. The Slick documentation states that it is possible to access even not directly supported databases, with a limited feature set, but I cannot see a way, how to accomplish that.

The forConfig-call throws the exception
`Configured profile com.databricks.client.jdbc.Driver does not conform to requested profile slick.jdbc.JdbcProfile`

**Comment by nafg** [comment]
Can you share a reproducible setup?

**Reply by hbergmey** [reply]
Unfortunately the Databricks server is a protected cloud instance of our customer. But as far as I can see, the exception flies before the connection attempt, So this is the conf;
```
    db {
      driver = "com.databricks.client.jdbc.Driver"
      url = "jdbc:databricks://SOME_IP:443;HttpPath=/sql/1.0/endpoints/abcd;TransportMode=http;SSL=1"
      PWD = "xxxxxx"
      UID = "token"
      AuthMech = "3"
    }
```

build.sbt
```
libraryDependencies` ++= Seq(
   "com.databricks" % "databricks-jdbc" % "2.6.29",
   "com.typesafe.slick" %% "slick" % "3.3.3"
)
```

And here a small test:
```
  def main(args : Array[String]) = {
    implicit val ec : ExecutionContext = ExecutionContext.global
    val slickSession = SlickSession.forConfig("db")
    import slickSession.profile.api._
    val db = slickSession.db

    val query = sql"SELECT 1".as[Int]
    val result = Await.result(db.run(query), 10.seconds)
    println(s"""Result of "SELECT 1" is `$result`""")
  }
```

Here is the Databricks JDBC documentation: https://docs.databricks.com/integrations/bi/jdbc-odbc-bi.html#jdbc-driver.

As far as I understand the documentation, Databricks understands ANSI SQL-92. Maybe I can "absuse" another profile?

**Reply by nafg** [reply]
What is SlickSession?

**Reply by hbergmey** [reply]
`akka.stream.alpakka.slick.scaladsl.SlickSession`. Oh d*mn, don't tell me, it is outdated. You caught me forgetting to mention my Alpakka dependency:
```
"com.lightbend.akka"    %% "akka-stream-alpakka-slick" % "3.0.4"
```

The Slick documentation describes `Database.forConfig` but I do not know, how to import `Database`. I suppose it would be provided by the correct JdbcProfile implementation, but this is exactly what I am missing for a not immediately supported database kind. This is why I relied on Alpakka SlickSession's auto-detection, which apparently does not work.

So, is there anything like an ANSISql92Profile, or a GenericJdbcProfile?

**Reply by nafg** [reply]
I'm not sure what you mean exactly but it sounds like the issue does not require a databricks server to reproduce, so if you could create a git repo or branch that I can check out, do `sbt run`, and see the issue, I can try to help G-d willing.

**Reply by nafg** [reply]
Also if I do need an actual server is there a way I could get one, perhaps you could set up a different one for me to test against.

Alternatively, if your customer wants to pay for me to help, we can take this to a private conversation and I would be more available sooner, and that might resolve the server access problem.

**Reply by hbergmey** [reply]
Here you are. https://github.com/hbergmey/generic-jdbc-connection-test Just clone and `sbt run`.
Keep in mind, that the problem is not Databricks specific and that the error occurs long before an actual connection is even attempted.

My real problem is, that Slick claims to be able to connect to not directly supported databases, but does not give an example of how to actually accomplish that.

Quoted from: https://scala-slick.org/doc/3.3.3/supported-databases.html:

> Supported Databases
> 
>     DB2
>     Derby
>     H2
>     HSQLDB
>     Microsoft SQL Server
>     MySQL
>     Oracle
>     PostgreSQL
>     SQLite
> 
>  Other SQL databases can be accessed right away with a reduced feature set. 

In all cases I find, I have to specify a "profile" explicitly, but there is no generic ANSI SQL profile. At least, I am unable to find it, nor do I have the slightest clue of how to implement that.

At the moment I'd be happy with a "reduced feature set", be able to formulate plain sql selects (https://scala-slick.org/doc/3.3.3/sql.html) and just use result streaming of rows.

My current concrete situation is related to a Databricks connection, but that is neither relevant nor could I provide an instance. Unfortunately Databricks is a cloud only provider with trial licenses and so on.

**Comment by nafg** [comment, marked as answer]
Ok so it seems like the confusion is about the term "driver," which may be partially the fault of the docs. "Driver" usually means the JDBC library that speaks a given database's wire protocol. However, earlier versions of Slick also referred to the database-specific Slick trait as a driver. Later, the term for that was changed to "profile."

If https://scala-slick.org/doc/stable/database.html is confusing or incorrect a PR to fix it would be appreciated. In any case, the more up to date specification of the config format is the API docs: https://scala-slick.org/doc/3.4.0/api/slick/jdbc/JdbcBackend$DatabaseFactoryDef.html#forConfig(String,Config,Driver,ClassLoader):Database

Bottom line, the `driver` config entry was interpreted by Slick as the deprecated way of specifying the Slick profile. Instead it should use the `profile` key. In any case it needs you to subclass `JdbcProfile` and give it the class name (suffixed with `$` if it's a Scala singleton object).

I sent a fix in https://github.com/hbergmey/generic-jdbc-connection-test/pull/1. The only error I get is  a timeout, I assume because it's trying to connect to `SOME_IP`.

**Reply by hbergmey** [reply]
Ok, this appears to work. It required another change to the configuration. I wrapped the authentication parameters in a "properties"-Block like so and moved MyProfile to a package `db`:
```
myGenericJdbcConfig {
  profile = "db.MyProfile$"
  db {
    driver = "com.databricks.client.jdbc.Driver"
    url = "jdbc:databricks://SOME_IP:443;HttpPath=/sql/1.0/endpoints/abcd;TransportMode=http;SSL=1"
    properties = {
      PWD = "xxxxxx"
      UID = "token"
      AuthMech = "3"
    }
  }
}
```

Then it can be loaded as `SlickSession.fromConfig("myGenericJsbcConfig")`

Now, two things. I am not fully aware, what set of features such an empty MyProfile delivers and which not. And I am missing a documentation chapter about implementing profiles.

The concepts page does not explain much: https://scala-slick.org/doc/stable/concepts.html.
And the API documentation about the JdbcProfile trait is kind of overwhelming: https://scala-slick.org/doc/3.4.0/api/slick/jdbc/JdbcProfile.html without explaining much. I have had a look into the Postgres-Implementation but confess, that that overwhelmed me, too.

**Reply by nafg** [reply]
What dialect is it similar to?

**Reply by hbergmey** [reply]
And once more I am the miserable unfortunate who is not of much help. All I know about the Databricks SQL is that it claims to be ANSI SQL Standard.
https://www.databricks.com/blog/2021/11/16/evolution-of-the-sql-language-at-databricks-ansi-standard-by-default-and-easier-migrations-from-data-warehouses.html

If I now knew, that any of the others is close to ANSI SQL, I'd be glad. Does the following type comparison provide any insight?
https://stackoverflow.com/questions/69123845/type-system-ansi-sql-standard-vs-postgres-oracle-db2

**Reply by hbergmey** [reply]
I am still wondering, what would be the best place to document this. First I thought, in the "Using Typesafe Config" as a third example, but then again, one has to specify a profile directly. So I guess it should be following DatabaseConfig. https://scala-slick.org/doc/stable/database.html#databaseconfig. And then a reference from the Supported Databases page would be good.

I'll give it a shot. I hope, my English is not too terrible.

**Reply by hbergmey** [reply]
## Generic JDBC or Custom Profile
In case You have to configure a connection to a not immediately supported type of database, you are required to provide a `JdbcProfile` for the database's SQL dialect in addition to spcifying the JDBC driver. This minimal example does not provide any custom dialect logic, but instead simply provides the default feature set of the `slick.jdbc.JdbcProfile` by extending it by an empty singleton object.

```scala
package org.custom.jdbc

import slick.jdbc.JdbcProfile

object MyDialectProfile extends JdbcProfile
```

The profile can then be used in the configuration.
```properties
my_other_db {
  profile = "org.custom.jdbc.MyDialectProfile$"
  db {
    url = "jdbc:other:db:url"              // refer to JDBC driver documentation for its specific URL syntax
    driver = org.some.custom.jdbc.Driver   // the actual java.sql.Driver implementation
    [...]                                  // further properties for the connection pool or specific to the driver
  }
}
```

And the configuration can be loaded as mentioned before using
```
val otherDc = DatabaseConfig.forConfig[JdbcProfile]("my_other_db")
```

**Reply by hbergmey** [reply]
Yes, ok, my attempt is a little clumsy, but I would like to have it small and simple. What is oviously missing here: What is the provided default feature set? From the source of JdbcProfile I infer that at least the basic column and query types should be working.

Is there any particular reason, why anything like `object GenericJdbcProfile extends JdbcProfile` does not already exist in Slick? Or do I happen to be I the first man on Mars? Is there a guide to implementing custom JdbcProfile? What baffles me, is that JdbcProfile is trait, but it does not specify any abstract methods or fields.

**Reply by nafg** [reply]
I think it existed at one point. I don't know. I'm trying to get Slick back on track but it has a long history before me.

Anyway send a PR and I can give feedback there.

BTW even with "concrete" profiles it's common to extend them as your own object. So having to extend JdbcProfile yourself is not such a relative burden.

**Reply by hbergmey** [reply]
<sigh>, now I first have to get the CLA signed by my superior, setup the project, fork it, figure out, how the docs are built... make buereaucracy happy, for three helpful sentences, that could just be copied in there.

I don't want to be ungrateful for your help, so I will see, what and when I can do that. It's just... we could both be working on something else already and still have the info available for others.

**Reply by nafg** [reply]
Just use the GitHub web UI

**Reply by nafg** [reply]
Just go to the file in github, click the edit button, follow the prompts. It's super simple

**Reply by ElectricWound** [reply]
Sorry, it has been a terribly busy month. I have contributed a PR using my private account.
